File: WebsocketServer/WebsocketServer/TypeScriptClient/lib/utilities/TreeItem.js
var WebsocketTSClient;
(function (WebsocketTSClient) {
    /**
     * 可以携带信息的树状数据结构，一对多结构
     */
    var TreeItem = /** @class */ (function () {
        function TreeItem(data) {
            /**
             * 携带的信息
             */
            this.data = null;
            /**
             * Id 由 Collection 创建，在 Collection 保持唯一
             */
            this.id = 0;
            this.parentId = null;
            this.childrenIds = [];
            this.data = data;
        }
        /**
         * to json for network transport, 携带的信息继承了 ITreeItem，也可以一并传输
         */
        TreeItem.prototype.toJson = function () {
            var jobj = {
                Id: this.id,
                ParentId: this.parentId,
                ChildrenIds: this.childrenIds,
                data: this.data
            };
            return jobj;
        };
        return TreeItem;
    }());
    WebsocketTSClient.TreeItem = TreeItem;
    (function (TreeItem) {
        /**
        * 负责管理树状数据结构
        */
        var Collection = /** @class */ (function () {
            function Collection() {
                /**
                 * flat list tree item，用于网络传输与检索
                 */
                this.items = [];
                this.topMostItems = [];
            }
            /**
             * 获取指定 ID 的节点
             */
            Collection.prototype.getItemById = function (id) {
                if (id >= 0 && id < this.items.length) {
                    return this.items[id];
                }
                return null;
            };
            /**
             * 获取所有节点
             */
            Collection.prototype.getAllItems = function () {
                return this.items;
            };
            /**
             * 获取所有顶级节点
             */
            Collection.prototype.getTopMostItems = function () {
                return this.topMostItems;
            };
            /**
             * Safe remove item, if the item has any children, it means it's a dependency node, it can not be removed until the children are clear
             */
            Collection.prototype.safeRemoveItem = function (id) {
                var itemToRemove = this.items.find(function (item) { return item.id === id; });
                if (!itemToRemove) {
                    return false; // Item with the given ID not found
                }
                if (itemToRemove.childrenIds.length > 0) {
                    return false; // Item has children, cannot be removed
                }
                return this.removeItem(id);
            };
            /**
             * Remove an item from the collection by its ID
             */
            Collection.prototype.removeItem = function (id) {
                var itemToRemove = this.items.find(function (item) { return item.id === id; });
                if (itemToRemove) {
                    if (itemToRemove.parentId === null) {
                        this.topMostItems = this.topMostItems.filter(function (item) { return item.id !== id; });
                    }
                    var removedIndex = this.items.indexOf(itemToRemove);
                    // Remove the item from its parent's ChildrenIds list
                    if (itemToRemove.parentId !== null) {
                        var parent_1 = this.items.find(function (item) { return item.id === itemToRemove.parentId; });
                        if (parent_1) {
                            parent_1.childrenIds = parent_1.childrenIds.filter(function (childId) { return childId !== id; });
                        }
                    }
                    // Remove the item from the flat list
                    this.items.splice(removedIndex, 1);
                    // Adjust IDs of items after the removed item
                    for (var i = removedIndex; i < this.items.length; i++) {
                        this.items[i].id--;
                    }
                    return true;
                }
                return false;
            };
            /**
             * 创建根节点
             */
            Collection.prototype.createRootItem = function (data) {
                return this.createItem(data, null);
            };
            /**
             * 创建新的树状数据结构节点
             */
            Collection.prototype.createItem = function (data, parent) {
                var item = new TreeItem(data);
                item.parentId = parent ? parent.id : null;
                item.id = this.items.length;
                if (parent) {
                    parent.childrenIds.push(item.id);
                }
                else {
                    // 没有 parent 就是顶级节点
                    this.topMostItems.push(item);
                }
                this.items.push(item);
                return item;
            };
            /**
             * 从根节点开始操作，parent = null
             */
            Collection.prototype.startFromRoot = function () {
                return new Result(null, null, this);
            };
            /**
             * 从某个节点开始操作
             */
            Collection.prototype.startFrom = function (item) {
                return new Result(null, item, this);
            };
            Collection.prototype.toJson = function () {
                var arr = [];
                this.items.forEach(function (item) {
                    arr.push(item.toJson());
                });
                return arr;
            };
            Collection.parse = function (json) {
                var collection = new TreeItem.Collection();
                if (Array.isArray(json)) {
                    json.forEach(function (itemToken) {
                        var id = itemToken['Id'];
                        var parentId = itemToken['ParentId'];
                        var childrenIds = itemToken['ChildrenIds'] || [];
                        var data = itemToken['data'];
                        var treeItem = new TreeItem(data);
                        treeItem.id = id;
                        treeItem.parentId = parentId;
                        treeItem.childrenIds = childrenIds;
                        collection.items.push(treeItem);
                    });
                }
                return collection;
            };
            return Collection;
        }());
        TreeItem.Collection = Collection;
        ;
        var TopmostResult = /** @class */ (function () {
            function TopmostResult(collection) {
                this.collection = collection;
            }
            /**
             * 设置没有 parent 的顶层节点
             */
            TopmostResult.prototype.next = function (data) {
                var topMostItem = this.collection.createItem(data, null);
                return new Result(topMostItem, null, this.collection);
            };
            return TopmostResult;
        }());
        TreeItem.TopmostResult = TopmostResult;
        ;
        var Result = /** @class */ (function () {
            function Result(item, parent, collection) {
                this.parent = null;
                this.item = null;
                this.item = item;
                this.parent = parent;
                this.collection = collection;
            }
            /**
             * 设置子节点
             */
            Result.prototype.children = function (action) {
                action === null || action === void 0 ? void 0 : action(this.collection.startFrom(this.item));
                return this;
            };
            /**
             * 设置平级节点
             */
            Result.prototype.next = function (data) {
                var nextItem = this.collection.createItem(data, this.parent);
                return new Result(nextItem, this.parent, this.collection);
            };
            return Result;
        }());
        TreeItem.Result = Result;
        ;
    })(TreeItem = WebsocketTSClient.TreeItem || (WebsocketTSClient.TreeItem = {}));
})(WebsocketTSClient || (WebsocketTSClient = {}));

File: WebSocketUnityClient/Assets/WebSocketClient/Runtime/Scripts/Utilities/Data/TreeItem.cs
using Newtonsoft.Json.Linq;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using UIFramework;
using UnityEngine;
using UnityEngine.Assertions;

namespace WebSocketClient.Utilities.Data
{
    public interface INetworkTransport
    {
        JToken ToJson();
    }

    /// <summary>
    /// 可以携带信息的树状数据结构,一对多结构
    /// </summary>
    /// <typeparam name="T"></typeparam>
    public class TreeItem<T>
    {
        public T data;
        /// <summary>
        /// 不允许外部创建，只能继承类和 Collection 创建
        /// </summary>
        protected TreeItem(T data)
        {
            this.data = data;
        }

        /// <summary>
        /// Id 由 Collection 创建，在 Collection 保持唯一
        /// </summary>
        public int Id { get; private set; }
        public int? ParentId { get; private set; }
        public List<int> ChildrenIds { get; private set; } = new List<int>();

        public int Depth { get; private set; } = 0;

        /// <summary>
        /// to json for network transport,携带的信息继承了 ITreeItem， 也可以一并传输
        /// </summary>
        /// <returns></returns>
        public JToken ToJson()
        {
            JObject jobj = new JObject();
            jobj.Add("Id", Id);
            if (ParentId.HasValue)
            {
                jobj.Add("ParentId", ParentId);
            }
            jobj.Add("ChildrenIds", JHelper.MakeIntArray(ChildrenIds));           
            // 如果对象继承了ToJson接口，将携带的信息也传输出去
            if (typeof(T).GetInterfaces().Contains(typeof(INetworkTransport)))
            {
                jobj.Add("data", (data as INetworkTransport).ToJson());
            }
            return jobj;
        }

        /// <summary>
        /// 负责管理树状数据结构
        /// </summary>
        public class Collection
        {
            /// <summary>
            /// flat list tree item，用于网络传输与检索
            /// </summary>
            private readonly List<TreeItem<T>> items = new List<TreeItem<T>>();
           
            /// <summary>
            /// 获取指定 ID 的节点
            /// </summary>
            /// <param name="id">节点 ID</param>
            /// <returns>节点</returns>
            public TreeItem<T> GetItemById(int id)
            {
                if (id >= 0 && id < items.Count)
                {
                    return items[id];
                }
                return null;
            }

            /// <summary>
            /// 获取所有节点
            /// </summary>
            /// <returns>所有节点</returns>
            public IEnumerable<TreeItem<T>> GetAllItems()
            {
                return items;
            }

            /// <summary>
            /// 获取所有顶级节点
            /// </summary>
            /// <returns></returns>
            public IEnumerable<TreeItem<T>> GetTopMostItems()
            {
                return items.Where(p => p.Depth == 0);
            }        

            /// <summary>
            /// Safe remove item, if the item has any children, it means it's a dependency node, it can not be removed until the children is clear
            /// </summary>
            /// <param name="id">The ID of the item to remove</param>
            /// <returns>True if the item was removed successfully, otherwise false</returns>
            public bool SafeRemoveItem(int id)
            {
                TreeItem<T> itemToRemove = items.FirstOrDefault(item => item.Id == id);
                if (itemToRemove == null)
                {
                    return false; // Item with the given ID not found
                }

                if (itemToRemove.ChildrenIds.Count > 0)
                {
                    return false; // Item has children, cannot be removed
                }
                return RemoveItem(id);
            }

            /// <summary>
            /// Remove an item from the collection by its ID
            /// </summary>
            /// <param name="id">The ID of the item to remove</param>
            /// <returns>True if the item was removed successfully, otherwise false</returns>
            public bool RemoveItem(int id)
            {       

                TreeItem<T> itemToRemove = items.FirstOrDefault(item => item.Id == id);
                if (itemToRemove != null)
                {                    
                    int removedIndex = items.IndexOf(itemToRemove);

                    // Remove the item from its parent's ChildrenIds list
                    if (itemToRemove.ParentId.HasValue)
                    {
                        TreeItem<T> parent = items.FirstOrDefault(item => item.Id == itemToRemove.ParentId);
                        if (parent != null)
                        {
                            parent.ChildrenIds.Remove(id);
                        }
                    }

                    // Remove the item from the flat list
                    items.Remove(itemToRemove);

                    // Adjust IDs of items after the removed item
                    for (int i = removedIndex; i < items.Count; i++)
                    {
                        items[i].Id--;
                    }

                    return true;
                }
                return false;
            }

            /// <summary>
            /// 创建根节点
            /// </summary>
            /// <returns>新节点</returns>
            public TreeItem<T> CreateRootItem(T data)
            {
                return CreateItem(data, null);
            }

            /// <summary>
            /// 创建新的树状数据结构节点
            /// </summary>
            /// <returns>新节点</returns>
            public TreeItem<T> CreateItem(T data, TreeItem<T> parent)
            {
                var item = new TreeItem<T>(data)
                {
                    ParentId = parent?.Id,
                    Id = items.Count
                };

                if (parent != null)
                {
                    item.Depth = parent.Depth + 1;
                    parent.ChildrenIds.Add(item.Id);
                }
                else 
                {
                    /// 没有 parent 就是顶级节点
                    item.Depth = 0;
                }

                items.Add(item);
                return item;
            }

            /// <summary>
            /// 从根节点开始操作，parent = null
            /// </summary>
            /// <returns></returns>
            public Result StartFromRoot()
            {
                return new Result(null, null, this);
            }

            /// <summary>
            /// 从某个节点开始操作
            /// </summary>
            /// <returns></returns>
            public Result StartFrom(TreeItem<T> item)
            {
                return new Result(null, item, this);
            }

            public JToken ToJson()
            {
                JArray arr = new JArray();
                foreach (var item in items)
                {
                    arr.Add(item.ToJson().ToString(Newtonsoft.Json.Formatting.None));
                }
                return arr;
            }

            public static Collection Parse(JToken json)
            {
                Collection collection = new Collection();

                if (json is JArray array)
                {
                    foreach (var jstr in array)
                    {
                        var itemToken = JObject.Parse(jstr.ToString());
                        int id = JHelper.GetJsonInt(itemToken, "Id");
                        int? parentId = null;
                        if (itemToken.ContainsKey("ParentId"))
                        {
                            parentId = JHelper.GetJsonInt(itemToken, "ParentId");
                        }
                        List<int> childrenIds = JHelper.GetJsonIntArray(itemToken, "ChildrenIds").ToList();
                        // Since the data might be an object implementing INetworkTransport, we'll need to deserialize it
                        
                        var jData = JHelper.GetJsonToken(itemToken, "data");
                        T data = jData.ToObject<T>();

                        // Create the tree item
                        TreeItem<T> treeItem = new TreeItem<T>(data)
                        {
                            Id = id,
                            ParentId = parentId,
                            ChildrenIds = childrenIds
                        };

                        // Add the tree item to the collection
                        collection.items.Add(treeItem);
                    }
                }
                return collection;
            }
        }

        public class TopmostResult
        {
            private Collection collection;

            public TopmostResult(Collection collection)
            {
                this.collection = collection;
            }

            /// <summary>
            /// 设置没有 parent 的顶层节点
            /// </summary>
            public Result Next(T data)
            {
                var topMostItem = collection.CreateItem(data, null);
                return new Result(topMostItem, null, collection);
            }
        }

        public class Result
        {
            public TreeItem<T> parent { get; private set; }
            public TreeItem<T> item { get; private set; }

            private Collection collection;

            public Result(TreeItem<T> item, TreeItem<T> parent, Collection collection)
            {
                this.item = item;
                this.parent = parent;
                this.collection = collection;
            }                      

            /// <summary>
            /// 设置子节点
            /// </summary>
            public Result Children(Action<Result> action)
            {
                action?.Invoke(collection.StartFrom(item));                
                return this;
            }

            /// <summary>
            /// 设置平级节点
            /// </summary>
            public Result Next(T data)
            {
                var nextItem = collection.CreateItem(data, parent);
                return new Result(nextItem, parent, collection);
            }
        }
    }
}
